sktime/annotation/ggs.py

```diff
@@ -0,0 +1,320 @@
+        """
+        change_points = self.initialize_change_points(data)
+        self._intermediate_change_points = [change_points[:]]
+        self._intermediate_ll = [self.cumulative_log_likelihood(data, change_points)]
+
+        # Start GGS Algorithm
+        for _ in range(self.k_max):
+            new_index = -1
+            new_value = +1
+            # For each segment, find change point and increase in LL
+            for start, stop in zip(change_points[:-1], change_points[1:]):
+                segment = data[start:stop, :]
+                ind, val = self.add_new_change_point(segment)
+                if val < new_value:
+                    new_index = ind + start
+                    new_value = val
+
+            # Check if our algorithm is finished
+            if new_value == 0:
+                logger.info("Adding change points!")
+                return change_points
+
+            # Add new change point
+            change_points.append(new_index)
+            change_points.sort()
+            if self.verbose:
+                logger.info(f"Change point occurs at: {new_index}, LL: {new_value}")
+
+            # Adjust current locations of the change points
+            change_points = self.adjust_change_points(data, change_points, [new_index])[
+                :
+            ]
+
+            # Calculate likelihood
+            ll = self.cumulative_log_likelihood(data, change_points)
+            self._intermediate_change_points.append(change_points[:])
+            self._intermediate_ll.append(ll)
+
+        return change_points
+
+    def predict(self, X: npt.ArrayLike):
+        """Predict."""
+        self.change_points_ = self.find_change_points(X)
+
+        labels = np.zeros(X.shape[0], dtype=np.int32)
+        for i, (start, stop) in enumerate(
+            zip(self.change_points_[:-1], self.change_points_[1:])
+        ):
+            labels[start:stop] = i
+        return labels
+
+
+class GGSEstimator:
+    """Sklearn Adapter."""
+
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
+        self._adaptee_class = GGS
+        self._adaptee = self._adaptee_class(**self.kwargs)
+
+    def fit(self, X: npt.ArrayLike, y: npt.ArrayLike = None):
+        """Fit."""
+        self._adaptee.initialize_intermediates()
+        return self
+
+    def predict(self, X: npt.ArrayLike, y: npt.ArrayLike = None) -> npt.ArrayLike:
+        """Predict."""
+        return self._adaptee.predict(X)
+
+    def fit_predict(self, X: npt.ArrayLike, y: npt.ArrayLike = None) -> npt.ArrayLike:
+        """Fit and predict."""
+        return self.fit(X, y).predict(X, y)
+
+    def get_params(self) -> Dict:
+        """Return initialization parameters."""
+        return asdict(self._adaptee, filter=lambda attr, value: attr.init is True)
+
+    def set_params(self, *args, **kwargs):
+        """Initialize the estimator from parameters."""
+        self._adaptee = self._adaptee_class(*args, **kwargs)
```

Comment:
I slightly worried about doing `set_params` this way for 2 reasons:
- by creating a new estimator rather than changing the previous one I wonder if you might create weird bugs for code that needs the same object instance in several places. (This might be an edge case, but an example of this came up recently in pywatts where a pipeline that uses one estimator as a transformer in one place, and then wishes to use the inverse version of that same object instance later). 
- if you do want to do it this way I think you need to copy over all the other params that have been set, whereas I think with this implementation you only instantiate with the 'new' parameters.

Reply:
agreed, likely to cause problems due to teh above.

Reply:
I mean overall I think it is nice to be consistent with the way `sklearn` does it, where the object is modified rather than a new object is created. 

I agree though, concrete examples are always a good idea.  (And perhaps I should also say that both points were probably more targeted at a general `BaseObject`-like thing with `set_params`, rather than GGS in particular).  Along those lines, here are some general examples that might work for both points:

1) I think the first example is in line with what I mentioned before - ie we use our estimator as a transformer, call `set_params` on it and then want to use the inverse transform of the same object we did the original transformation with (this is important if there were fitted params that play a role in both `transform` and `inverse_transform`).  Could test this by having such an transformer and making sure you get the same data back if you `transform`, `set_params` and then `inverse_transform`.  
    - If this is something that is actually likely to happen? Perhaps not? Perhaps it is worth having a discussion as to what the expectations are here. But again, don't think defaulting to what `sklearn` does is such a bad idea. 
2) - If I first use set_params to set attribute a, then use it to set attribute `b` I would expect the final estimator that is returned to have both the `a` and the `b` estimator updated.  Another think to potentially check is that any attributes we set in init are correctly maintained in the object returned if they were not the attribute being changed via `set_params`.

Personally I wouldn't say that you need to implement either test here (unless you strongly want to!) as I think they won't really be as relevant to what is going on in GGS.  And I think your fix is now consistent.  But think it could be useful to do for the more general `BaseObject`/`BaseEstimator`. 

Think we can improve on the `set_params` etc in the more general `BaseObject` PR! :)  But figure its not worth going too much into the details on GGS.

Reply:
> I'm am not necessarily a fan partially modifying the internal state of an estimator - since it leaves it essentially in a corrupted state where __init__ parameters do not correspond to the attributes set at fit. I though that cutting a fresh estimator may prevent problems rather than create but probably a concrete case would help to reduce the uncertainty here.

I can see your point here though as well. I *think* the interpretation via `sklearn` is that `set_params` is meant to be the approved interface for changing the `__init__` parameters, so that this is ok! But it is an assumption.  Again I think it boils down to - does the user expect a clean state when calling `set_params`?

Reply:
Thanks for providing a specific example but I am not sure how relevant it is since this estimator is not a transformer and therefore it has no `inverse_transform` method. Might be worth adding this to the base class discussion.